src/geode/geometry/nn_search.cpp

```diff
@@ -190,14 +210,27 @@ namespace geode
         typename NNSearch< dimension >::ColocatedInfo result;
         std::vector< index_t > mapping( nb_points() );
         absl::c_iota( mapping, 0 );
+        std::mutex mutex;
         async::parallel_for( async::irange( index_t{ 0 }, nb_points() ),
-            [&epsilon, &mapping, this]( index_t p ) {
-                if( mapping[p] == p )
+            [&epsilon, &mapping, &mutex, this]( index_t p ) {
+                if( mapping[p] != p )
+                {
+                    return;
+                }
+                const auto vertices = radius_neighbors( point( p ), epsilon );
+                auto min_index = find_min_unmapped_element( vertices, mapping );
```

Comment:
not sure to see the difference with min_element(vertices)

Reply:
considering only element == mapping[element]
it is the crucial change of this PR